server/rpm/pbench-server.spec.j2

```diff
@@ -58,8 +62,11 @@ cp -a ./web-server/* %{buildroot}/%{installdir}/%{static}
 mv %{buildroot}/%{installdir}/%{static}/package.json %{buildroot}/%{installdir}
 
 %post
-# Install python dependencies
-pip3 --no-cache-dir install --prefix=/%{installdir} -r /%{installdir}/requirements.txt > /%{installdir}/pip3-install.log
+# Install python dependencies for pbench user
+(su pbench -c "pip3 --no-cache-dir install --user -r /%{installdir}/requirements.txt") > /%{installdir}/pip3-install.log
+
+# Add install directory into the pbench user default PYTHONPATH
+su pbench -c "echo /%{installdir}/lib > \$(python3 -m site --user-site)/pbench.pth"
```

Comment:
What adds `pbench.pth` to the path?  Could this commit be missing a file change?

Reply:
That's a feature of the `site` package, which is what sets up the initial default `sys.path`.

I can add to the comment, though, to clarify this.

Reply:
So by running `python3 -m site --user-site` it creates `pbench.pth`?

It looks like `python3 -m site --user-site` emits a path to which `pbench.pth` is appended, and then the `echo /${installdir}/lib` is creating that file.

But who reads `pbench.pth`?

Reply:
OK. Yeah, this is `site` magic, and it took me a while to work this out. Basically, the `site` package is implicit Python magic; it's run automatically when Python starts, and sets up `sys.path` on top of the external `PYTHONPATH`.

For our purposes, the `site` magic includes looking in the per-user site library and adding the `site-packages` paths *and* adding any directories captured in `*.pth` files in the root directory. This is pretty cool, if more than a bit obscure.

So the line here is taking our install path `/${installdir}/lib` and storing that path into our own little `pbench.pth` file for `site` to find when the pbench user runs Python.

That `pbench.pth` file needs to be in the right place, and the `python3 -m site --user-site` gives us the proper directory where it'll look. The `su pbench` ensures that this is run *as* "pbench" and therefore gives us the proper `/home/pbench/.local/lib/python3.6/site-packages` path where that file belongs.

Now, when the pbench user runs Python, *without* any `PYTHONPATH` environment variable, and asks for `sys.path`, it'll include the pbench user site-library paths, *plus* `/opt/pbench-server/lib`.

Maybe I over-complicated the example above, because I asked the site package to give me its nice formatting of `sys.path`. Let's try this instead:

```
[pbench@dbutenho-dev root]$ echo $PYTHONPATH

[pbench@dbutenho-dev root]$ whoami
pbench
[pbench@dbutenho-dev root]$ ls $(python3 -m site --user-site)
alembic                    botocore                    elasticsearch-7.13.1.dist-info   gunicorn-20.1.0.dist-info        s3transfer-0.5.0.dist-info
alembic-1.6.5.dist-info    botocore-1.21.28.dist-info  email_validator                  humanize                         sqlalchemy
aniso8601                  click                       email_validator-1.1.3.dist-info  humanize-3.11.0.dist-info        SQLAlchemy-1.4.23.dist-info
aniso8601-9.0.1.dist-info  click-8.0.1.dist-info       flask                            pbench.pth
boto3                      dateutil                    Flask-2.0.1.dist-info            python_dateutil-2.8.2.dist-info
boto3-1.18.28.dist-info    elasticsearch               gunicorn                         s3transfer
[pbench@dbutenho-dev root]$ cat $(python3 -m site --user-site)/pbench.pth
/opt/pbench-server/lib
[pbench@dbutenho-dev root]$ python3
Python 3.6.8 (default, Mar 18 2021, 08:58:41) 
[GCC 8.4.1 20200928 (Red Hat 8.4.1-1)] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> import sys
>>> sys.path
['', '/usr/lib64/python36.zip', '/usr/lib64/python3.6', '/usr/lib64/python3.6/lib-dynload', '/home/pbench/.local/lib/python3.6/site-packages', '/opt/pbench-server/lib', '/usr/local/lib64/python3.6/site-packages', '/usr/local/lib/python3.6/site-packages', '/usr/lib64/python3.6/site-packages', '/usr/lib/python3.6/site-packages']
```

Reply:
Awesome!  That is a great explanation.  Would it be possible to place a comment in the `.spec.j2` file?

Reply:
Thanks!